language-plutus-core/src/Language/PlutusCore/Check/ValueRestriction.hs

```diff
@@ -0,0 +1,58 @@
+-- | The value restriction check.
+
+{-# LANGUAGE LambdaCase #-}
+
+module Language.PlutusCore.Check.ValueRestriction
+    ( checkValueRestriction
+    , checkValueRestrictionProgram
+    ) where
+
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Type
+import           Language.PlutusCore.View  (isTermValue)
+
+import           Control.Monad.Error.Lens
+import           Control.Monad.Except
+import           Data.Foldable             (traverse_)
+import           Data.Functor.Foldable
+
+{- Note [Value restriction]
+We need to report a value restriction violation in case a term that goes after a 'TyAbsF'
+is not a value. However there can be terms like
+
+    /\(a :: *) -> /\(b :: *) -> error {a}
+
+and we want to report that value restriction is violated after the @(b :: *)@ binding rather than
+after the @(a :: *)@ binding, even though
+
+    /\(b :: *) -> error {a}
+
+is already not a value and thus we could report "earlier", but that would be misleading, because
```

Comment:
I disagree with this: I claim that `/\(b :: *) -> error {a}` is just not well-formed. I think that means it's somewhat senseless to ask if it is a value or not. If anything I think we should say it *is* a value, because then our value check can be nice and stupid: type abstractions are always values.

That does mean we have an awkward pair of mutually recursive checks: is-well-formed and is-value. Choosing to treat non-well-formed type abstractions as values breaks one direction of the recursion, which seems good. This also has the property that it will remain correct if the type abstraction becomes well-formed, because well-formed type abstractions are always values.

Reply:
That's clever. Fixing.

Reply:
> That does mean we have an awkward pair of mutually recursive checks: is-well-formed and is-value.

I'm not sure how to make this work. So we can't check whether a term is a value without knowing whether the term is valid, but we could check both the things at the same time, but then we need to fail short if the body of a type abstraction is not a value, but then again, we can't check this without checking the validity first, so we end up traversing the entire program regardless of whether there is any value restriction violation or not.

> If anything I think we should say it is a value, because then our value check can be nice and stupid: type abstractions are always values.

This works, I think. We just ensure that bodies of type abstractions are values and later confirm that those bodies are indeed valid. This is efficient and meaningful as far I can tell. Implemented that.

Reply:
> This works, I think. We just ensure that bodies of type abstractions are values and later confirm that those bodies are indeed valid. This is efficient and meaningful as far I can tell. Implemented that.

:+1: